temporian/implementation/numpy/operators/cast.py

```diff
@@ -0,0 +1,105 @@
+from typing import Dict
+
+import numpy as np
+
+from temporian.core.data.dtype import DType
+from temporian.core.operators.cast import CastOperator
+from temporian.implementation.numpy.data.event import NumpyEvent
+from temporian.implementation.numpy.data.feature import (
+    NumpyFeature,
+    DTYPE_REVERSE_MAPPING,
+)
+from temporian.implementation.numpy import implementation_lib
+from temporian.implementation.numpy.operators.base import OperatorImplementation
+
+
+class CastNumpyImplementation(OperatorImplementation):
+    def __init__(self, operator: CastOperator) -> None:
+        super().__init__(operator)
+        self._dtype_limits = {
+            DType.INT32: np.iinfo(np.int32),
+            DType.INT64: np.iinfo(np.int64),
+            DType.FLOAT32: np.finfo(np.float32),
+            DType.FLOAT64: np.finfo(np.float64),
+        }
+
+        self._nocheck_dtypes = [DType.BOOLEAN, DType.STRING]
+
+    def _can_overflow(self, origin_dtype: DType, dst_dtype: DType) -> bool:
+        # Don't check overflow for BOOLEAN or STRING:
+        #  - boolean: makes no sense, everybody knows what to expect.
+        #  - string: on orig_dtype, too costly to convert to numeric dtype
+        #            and compare to the limit. On dst_type, there's no limit.
+        if (
+            origin_dtype in self._nocheck_dtypes
+            or dst_dtype in self._nocheck_dtypes
+        ):
+            return False
+        return (
+            self._dtype_limits[origin_dtype].max
+            > self._dtype_limits[dst_dtype].max
+        )
+
+    def _check_overflow(
+        self,
+        data: np.ndarray,
+        origin_dtype: DType,
+        dst_dtype: DType,
+    ) -> None:
+        if np.any(
+            (data < self._dtype_limits[dst_dtype].min)
+            | (data > self._dtype_limits[dst_dtype].max)
+        ):
+            raise ValueError(
+                f"Overflow casting {origin_dtype}->{dst_dtype} {data=}"
+            )
+
+    def __call__(self, event: NumpyEvent) -> Dict[str, NumpyEvent]:
+        target_dtypes = self.operator.attributes["target_dtypes"]
+        check = self.operator.attributes["check_overflow"]
+
+        # Reuse event if actually no features changed dtype
+        if all(
+            orig_dtype is target_dtypes[feature_name]
+            for feature_name, orig_dtype in event.dtypes.items()
+        ):
+            return {"event": event}
+
+        # Create new event, some features may be reused
+        output = NumpyEvent(data={}, sampling=event.sampling)
+
+        for feat_idx, feature_name in enumerate(event.feature_names()):
```

Comment:
Performance wise, it would be best to have the feature look inside of the index loop (instead of the opposite)

Reply:
Ok this is interesting, because I thought so too and got a bit obsessed with it at this point, because the benchmarks show otherwise, consistently.

So, the main thing with the proposed version, is that you need to call `_can_overflow()` at each event_idx/feature. So I made a dict with `can_overflow` results pre-calculated, and made sure to have only one `event_index` lookup per row in this new implementation.
But the benchmark still favored the current version.

Then I thought that maybe adding 10 features, having these repeated event_idx lookups per feature would favor the proposed version, but the difference is even larger in favor of this current version:
```
for feature, for event (current), with 10 features
----------------------------------------------------------------
cast(check=False):100                0.00036       0.00036
cast(check=True):100                 0.00097       0.00097
cast(check=False):1000000            0.00115       0.00115
cast(check=True):1000000             0.00370       0.00370
================================================================

for event, for feature (proposed), with 10 features:
----------------------------------------------------------------
cast(check=False):100                0.00172       0.00172
cast(check=True):100                 0.00241       0.00241
cast(check=False):1000000            0.00349       0.00349
cast(check=True):1000000             0.00662       0.00644
================================================================
```

In summary, it seems like being able to factor out all these small dict lookups out of the second loop (`from_features[feature_name]`, `event.dtypes[feature_name]`, reverse dtype map, and check overflow), speeds up the code despite having to lookup by `event_index` for each feature. Which is consistent with the O(1) of the hashmap lookup, but still seems anti-intuitive.

Any thoughts?

Reply:
Great analysis. Let's follow the benchmark :).

Possibly, you can add a TODO or NOTE for later re-visitation (i.e., maybe the EventV2 will change things).